### ISSUE TITLE
Feature Extraction refactoring discussion thread

### ISSUE THREAD

**Issue by pbailis** [stanford-futuredata/macrobase#94]
#### Old Pipeline: 
Ingest -> Scoping -> Outlier Detection -> Summarization

*Basic idea:* The outlier detector both:
  1. scores data
  2. marks points as outliers based on a percentile over scores (i.e., 99th percentile)

#### New Pipeline: 
Ingest -> Feature extraction -> Context selection -> Density estimation -> Summarization

The main change in this pipeline is the separation of Feature extraction from Outlier detection.

###### Feature Extraction:

*Basic idea:* performs arbitrary transformation on input data

*Input:* data point (metrics in R^N, attributes)
*Output:* data point (metrics in R^M, attributes)

###### Density Estimation

*Basic idea:* performs density estimation on the transformed attributes

*Input:* data point (metrics in R^M, attributes)
*Output:* data point with probability (metrics in R^M, attributes, probability density estimate)

##### Code changes required:
 1. Most existing outlier detectors become feature extraction methods.
 2. KDE becomes default density estimation technique; we can also convert the percentile estimators for backwards compatibility.
 3. Will need utility methods for manipulating metric points (see discussion below)

##### Some discussion points
 1. Do we want to update point metrics in-place? It may be beneficial to have the original metrics attached, but this may also be expensive, especially if we chain feature extraction. Maybe we make this a method that we can overload later.
 2. If we perform dimensionality reduction (e.g., streaming PCA), do we do it before or after scoping?
 3. What is our upgrade path for batch versus streaming? Do we want to maintain separate code paths? We don't have a streaming version of scoping or KDE right now, so do we deprecate the streaming execution engine for the time being?

**Comment by pbailis** [comment]
re: #3, all else considered, I am inclined to deprecate streaming for the time being.

**Comment by raininthesun** [comment]
> Do we want to update point metrics in-place? It may be beneficial to have the original metrics attached, but this may also be expensive, especially if we chain feature extraction. Maybe we make this a method that we can overload later.

If we are not using original metrics to detect outliers, then no need to keep it for later pipelines? A related problem is do we write the features on disk or keep them in memory to feed directly to later. I think we might want to write them to disk, if we need to execute the pipeline more than once; this is only if the feature extraction is relatively expensive compared to reading from disk. 



> If we perform dimensionality reduction (e.g., streaming PCA), do we do it before or after scoping?

I think before scoping. The dimensionality reduction is for the metric attributes, which are used for KDE. Scoping sees KDE as a blackbox. So whatever processing needed before running KDE should be done before running scoping.  From this point of view, maybe the pipeline should be

Ingest -> Feature extraction ->( dimensionality reduction )-> (Contextual )Density estimation -> Summarization


> What is our upgrade path for batch versus streaming? Do we want to maintain separate code paths? We don't have a streaming version of scoping or KDE right now, so do we deprecate the streaming execution engine for the time being?

I agree. From the above pipeline, streaming is essentially handled in the feature extraction phase right now.

**Comment by pbailis** [comment]
Notes from call:

Want to call the first stage `FeatureTransformation`!

Want to call DensityEstimation `OutlierClassification`

Think about:
 - How to support ensembling of classification methods...
 - Iterative pipelines
 - How to support provenance

Outcome: tuple-at-a-time dataflow interfaces, no batch anymore...

**Comment by pbailis** [comment]
Vivek's proposed lambda wrapper:
```
class MapTransform {
MapTransform(Function<Datum, Datum> t) {this.t = t}
Datum transform(Datum in) {return t(in);}
}

new MapTransform(x -> x * 2)
```

**Comment by mamikonyana** [comment]
I created #96 that's a try to create an API based on our discussions, please take a look and don't be shy on comments

**Comment by pbailis** [comment]
Closing since we've merged #96
